src/publicApi.ts

```diff
@@ -300,14 +301,31 @@ export interface WorkspaceData {
     phase: WorkspacePhase;
     description: string;
     lastUsed: Date;
-    recentFolders : string[];
+    recentFolders: string[];
 }
 
-export function rawWorkspaceToWorkspaceData(rawWorkspaces: Workspace): WorkspaceData;
+export function rawWorkspaceToWorkspaceData(rawWorkspaces: Workspace): WorkspaceData | undefined;
 export function rawWorkspaceToWorkspaceData(rawWorkspaces: Workspace[]): WorkspaceData[];
 export function rawWorkspaceToWorkspaceData(rawWorkspaces: Workspace | Workspace[]) {
     const toWorkspaceData = (ws: Workspace) => {
-        const url = new URL(ws.context!.contextUrl);
+        let url: URL;
+        try {
+            if (
+                ws.context?.details.case === 'git' &&
+                ws.context.details.value.normalizedContextUrl !== ws.context.contextUrl // backward compatible
+            ) {
+                url = new URL(ws.context.details.value.normalizedContextUrl);
+            } else {
+                const normalized = ContextURL.getNormalizedURL({ context: {}, contextURL: ws.context!.contextUrl } as any);
+                if (!normalized) {
+                    return undefined;
+                }
+                url = normalized;
+            }
+        } catch (e) {
+            // ignore
+            return undefined;
```

Comment:
it will be unexpected? probably better to log?

Reply:
I'd like to add, but it seems needs HostService\LogService\TelemetryService here and change lots of files so that we can send exception. Maybe we will need inject in repo? I'd like to keep it and add TODO as it's right now.

cc @jeanp413 for `Maybe we will need inject in repo?` this question

<img width="1418" alt="image" src="https://github.com/gitpod-io/gitpod-vscode-desktop/assets/20944364/77504daf-8d3e-46b6-9996-33aaa8d92c73">